subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java

```diff
@@ -679,12 +679,29 @@ public void rebuildTransitions(final List<SubscriptionBaseEvent> inputEvents, fi
 
         transitions = new LinkedList<SubscriptionBaseTransition>();
 
+        // DefaultSubscriptionDao#buildBundleSubscriptions may have added an out-of-order cancellation event (https://github.com/killbill/killbill/issues/897)
+        final SubscriptionBaseEvent cancellationEvent = Iterables.tryFind(inputEvents,
+                                                                          new Predicate<SubscriptionBaseEvent>() {
+                                                                              @Override
+                                                                              public boolean apply(final SubscriptionBaseEvent input) {
+                                                                                  return input.getType() == EventType.API_USER && ((ApiEvent) input).getApiEventType() == ApiEventType.CANCEL;
+                                                                              }
+                                                                          }).orNull();
```

Comment:
I am confused, i thought the logic in `filterOutDuplicateCancelEvents` was precisely to cleanup the stream and only keep the first cancel event -- and delete everything after this.

Reply:
As discussed:

* Will keep the code as-is for 0.18
* Will open a new PR to cleanup the code in 0.19